### ISSUE TITLE
Multi Agentic (We can select one agent at a time )

### ISSUE THREAD

**Issue by SyedMuqtasidAli** [All-Hands-AI/OpenHands#2563]
@mroch @jeremi @penberg @ohhmm 
Feature Request: Waterfall Model Workflow for Agents and multi agent
What problem or use case are you trying to solve?

Currently, in OpenDevin, users can select and activate one agent at a time, such as the SWE agent, Codeac Agent, or Planning Agent. This process requires manual intervention to switch from one agent to another, which can be inefficient and time-consuming. The need is to automate the workflow in a waterfall model where agents work sequentially without requiring manual activation for each step. This approach ensures that tasks are handled in a structured manner, from planning to execution to testing, improving the overall efficiency and effectiveness of the workflow.

Describe the UX of the solution you'd like

The user experience should be streamlined and intuitive:

Single Workflow Selection: Users can select a "Waterfall Model Workflow" option that initiates a predefined sequence of agent activities.
Automated Sequence: Once initiated, the Planning Agent starts first. Upon completion, the system automatically transitions to the Codeac Agent, followed by the Testing Agent.
Progress Tracking: Users should be able to monitor the progress of each phase through a dashboard or progress bar, showing the current agent in operation and estimated time to completion.
Notifications and Updates: Users receive notifications upon the completion of each phase, with summaries of what was accomplished and what the next steps are.
Interruption and Resumption: Users can pause the workflow if needed and resume from where it left off.
Do you have thoughts on the technical implementation?

The implementation could involve:

Workflow Controller: Develop a workflow controller that manages the sequence of agents. This controller handles the initiation, monitoring, and transition between different agents.
State Management: Implement a state management system to track the progress of each phase and manage transitions. This could be done using a state machine or workflow engine.
Inter-agent Communication: Ensure smooth communication between agents by defining clear interfaces and data exchange protocols.
UI/UX Enhancements: Update the user interface to include the workflow selection option, progress tracking dashboard, and notification system.
Error Handling: Implement robust error handling to manage and recover from failures at any stage in the workflow.
Configuration Files: Allow customization of the workflow sequence and parameters through configuration files or a user-friendly settings page.
Describe alternatives you've considered

Manual Sequential Execution: Users manually start each agent in sequence, but this approach lacks automation and can lead to inefficiencies and delays.
Parallel Execution: Running all agents in parallel could save time but may lead to resource conflicts and a less structured workflow, which might not be suitable for all projects.
Custom Workflows: Allow users to define custom workflows and sequences, offering flexibility but requiring more user input and configuration.
Additional context

Implementing a waterfall model workflow can significantly improve project management and execution efficiency in OpenDevin. By automating the transition between different phases, users can focus on higher-level tasks and decision-making. This feature aligns with best practices in software development and project management, ensuring a systematic approach to task execution.

**Comment by neubig** [comment]
Hi @SyedMuqtasidAli, thanks for the suggestion! This sounds pretty cool, but it'd be a big change. My suggestion would be to try to split it up into bite-sized pieces that could be implemented independently.

Just an idea, this might be a good sequencing:
- [ ] Take a look at PlannerAgent, test it out a bit locally, then evaluate it on SWE-Bench, and see if it can improve accuracy. We haven't carefully paid attention to PlannerAgent because we have just been using CodeAct, but I do think planning and plan management could potentially increase accuracy.
- [ ] Check whether PlannerAgent is appropriately visualized in the UI.
- [ ] Make it so that the agents used by PlannerAgent and the sequencing can be configured by the user (instead of hard-coded in the code).
- [ ] Expose this configuration to the UI somehow.
- [ ] Experiment with different types of sequencing on SWE-Bench, and set the sequencing that is the best at solving software engineering tasks to be the default for PlannerAgent.

If this sounds like a reasonable plan (I'd be happy for other suggestions), we could split them off into separate issues and work on them one-by-one.

**Comment by SyedMuqtasidAli** [comment]
@neubig Thanks for your response, but i want all agents to work sequentially like this, the planner agent makes plans and then passes this plan to codeact agent and then codeact agent writes code according to the planning and passes it to the testing agent like this. can open devin follow this ? as I am seeing every agent have their own prompts and they all are working according to the prompts.

**Comment by neubig** [comment]
Hey @SyedMuqtasidAli , yes, delegation is possible. See:
- DelegatorAgent: https://github.com/OpenDevin/OpenDevin/tree/main/agenthub/delegator_agent
- CodeActAgent delegates to BrowserAgent when it needs to web browse: https://github.com/OpenDevin/OpenDevin/tree/main/agenthub/codeact_agent

**Comment by SyedMuqtasidAli** [comment]
@neubig Thanks you very much sir, one more things, the planner agent class is used only for making plan but today I try to give a prompt to planner agent " Write a code for calculator in python " so planner agent first make plans and then it will generate a file with calculator code how an agent planner write code ? even I saw that its not delegate planning towards any other coder agent can you please clear this confusion how planner agent generating code?

**Comment by neubig** [comment]
Yeah, I think the planner agent only writes plans, so the delegator agent would first delegate to the planner agent, then delegate to the coding agent.

**Comment by SyedMuqtasidAli** [comment]
@neubig my question is, in the planner agent class code there is nothing mentioned of that the planner agent deligating the task towards code agent for writing code,

**Comment by neubig** [comment]
Sorry, just to clarify: it is DelegatorAgent that does delegation. PlannerAgent does not do delegation.

**Comment by SyedMuqtasidAli** [comment]
@neubig Thanks you for  the clarifications.

**Comment by assertion** [comment]
I'm also interested in multi-agent area, and I submit an issue in April: https://github.com/OpenDevin/OpenDevin/issues/1332

And while I acknowledge that we currently have multi-agent capabilities in OpenDevin, I must point out that they seem to be in their very early stages and are not yet performing well on specific coding tasks.

@neubig Do we have some plan or on-going discussion about the multi-agent framework, and maybe a big arch refactor to achieve this?

**Comment by assertion** [comment]
@tobitege 
Thank you for your insights and clarifications. I appreciate the practical perspective on the current state of agent capabilities.

I wanted to share that there has been significant academic discussion around the use of multi-agent systems for task decomposition and mutual verification to enhance output quality. These approaches have shown potential in improving the quality of work by leveraging the strengths of multiple agents in a collaborative manner.

Currently, it seems that OpenDevin’s primary focus is still largely on single-agent systems( mainly in codeact agent). The existing Delegate agent mechanism for multi-agent coordination appears to have seen limited advancements. Additionally, the micro agents being orchestrated seem to be overly simplistic, and the overall effectiveness of multi-agent chaining in practical scenarios is not very impressive.

What I hope to see in a more sophisticated system is a framework with multiple specialized agents, such as:

A planning agent that can decompose tasks and update them based on actual outputs.
An agent specifically for understanding repository contents.
A coding agent to handle the coding tasks.
A testing agent to manage test cases.
A review agent to oversee code quality and perform code reviews.
These agents would work collaboratively, sharing critical information and thereby ensuring a higher quality of output. Such a system would leverage the strengths of different agents working in tandem to efficiently and effectively complete complex tasks.

Looking forward to further discussions on how we can potentially incorporate these ideas into OpenDevin.

**Comment by neubig** [comment]
Hi @assertion , I don't think we have any concrete plans beyond what I wrote above. If you're interested you're welcome to try it out. I don't think we need a big architecture refactor to do this, we could largely build on top of the existing delegator agent and make it more advanced.

**Comment by tobitege** [comment]
Hello @assertion,

thank you for elaborating on the topic, I appreciate the details and discussion! 👍 

Just wanted to share a reference I found in our open PR's list, related to this topic, though I can't speak on the status of it:
[https://github.com/OpenDevin/OpenDevin/pull/2460]

The Issues page seems to have quite a few more suggestions, but implementation is a different topic. 😀

**Comment by PierrunoYT** [comment]
I also heard that multi agent can increase the output quality very much

Check: https://arxiv.org/abs/2406.04692

**Comment by ohhmm** [comment]
I am trying to make it useful and it is clear to me also that it is absolutely necessary for this to have different modes that are running using different specialized models to reach state that time&money spent is less with opendevin:
 BA -> Architect -> TL -> Dev <-> CI & QA

**Comment by neubig** [comment]
Thanks @ohhmm , any contributions are welcome in this direction!
